File: JavaScript/TypeORM/src/service/EntityService.js
import { Department } from "../entity/Department.js";
import { Person } from "../entity/Person.js";
import { Employee } from "../entity/Employee.js";
import { DataSource } from "typeorm/data-source/DataSource.js";

export class EntityService{
    /**
     *
     * @param {DataSource} connection
     */
    constructor(connection) {
        this.connection = connection
    }

    /**
     *
     * @param {number} id
     * @returns {Department | null}
     */
    async getDepartment(id) {
        const rep = this.connection.getRepository(Department.schemaName);
        /** @type {Department} */
        const dep = await rep.findOneBy({id: id});
        return new Department(dep.name)
    }

    /**
     *
     * @param {string} id
     * @returns {Person | null}
     */
    async getPerson(id) {
        const rep = this.connection.getRepository(Person.schemaName);
        /** @type {Person} */
        const per = await rep.findOneBy({id: id});
        return new Person(per.name, per.address, new Date(per.birthDate));
    }

    /**
     *
     * @param {string} id
     * @returns {Employee | null}
     */
    async getEmployee(id) {
        const rep = this.connection.getRepository(Employee.schemaName);
        /** @type {Employee} */
        const emp = await rep.findOne({
            where: {id: id},
            relations: {
                person: true,
                department: true,
            },
            relationLoadStrategy: "join",
        });
        return new Employee(emp.department, emp.person, emp.position);
    }

    /**
     *
     * @param {any} entity
     * @returns {Promise<any>}
     */
    #save(entity, schemaName) {
        let rep = this.connection.getRepository(schemaName);
        return rep.save(entity);
    }

    /**
     *
     * @param {int} name
     * @returns {Promise<Department>}
     */
    addNewDepartment(name) {
        let dep = new Department(name);
        return this.#save(dep, Department.schemaName);
    }

    /**
     *
     * @param {string} name
     * @param {string} address
     * @param {Date} birthDate
     * @returns {Promise<Person>}
     */
    addNewPerson(name, address, birthDate) {
        let per = new Person(name, address, birthDate);
        return this.#save(per, Person.schemaName);
    }

    /**
     *
     * @param {Person} person
     * @param {Department} department
     * @param {int} position
     * @returns {Promise<Employee>}
     */
    addNewEmployee(person, department, position) {
        let emp = new Employee(department, person, position);
        return this.#save(emp, Employee.schemaName);
    }

    /**
     *
     * @param {Department} department
     * @returns {Promise<Department>}
     */
    saveDepartment(department) {
        return this.#save(department, Department.schemaName);
    }

    /**
     *
     * @param {Person} person
     * @returns {Promise<Person>}
     */
    savePerson(person) {
        return this.#save(person, Person.schemaName);
    }

    /**
     *
     * @param {Employee} employee
     * @returns {Person<Employee>}
     */
    saveEmployee(employee) {
        return this.#save(employee, Employee.schemaName);
    }

    /**
     *
     * @param {Department} department
     * @param {string} name
     * @param {string} address
     * @param {Date} birthDate
     * @param {int} position
     * @returns
     */
    async addNewEmployeeAndPerson(department, name, address, birthDate, position) {
        let person = await this.addNewPerson(name, address, birthDate);
        return this.addNewEmployee(person, department, position);
    }

    /**
     *
     * @param {string} schemaName
     * @returns {Promise<any>}
     */
    #getAll(schemaName) {
        let rep = this.connection.getRepository(schemaName);
        return rep.find();
    }

    /**
     *
     * @returns {Department[]}
     */
    async getAllDepartments() {
        let entities = await this.#getAll(Department.schemaName);
        let departments = [];
        entities.forEach(e => {
           let dep = new Department(e.name);
           dep.id = e.id;
           departments.push(dep);
        });
        return departments;
    }

    /**
     *
     * @returns {Person[]}
     */
    async getAllPersons() {
        let entities = await this.#getAll(Person.schemaName);
        let persons = [];
        entities.forEach(e => {
            let per = new Person(e.name, e.address, new Date(e.birthDate));
            per.id = e.id;
            persons.push(per);
        });
        return persons;
    }

    /**
     *
     * @returns {Employee[]}
     */
    async getAllEmployees() {
        let rep = this.connection.getRepository(Employee.schemaName);
        let entities = await rep.find({
            relations: {
                person: true,
                department: true,
            },
            relationLoadStrategy: "join",
        });
        let employees = [];
        entities.forEach(e => {
            let dep = new Department(e.department.name);
            dep.id = e.department.id;
            let per = new Person(e.person.name, e.person.address, new Date(e.person.birthDate));
            per.id = e.person.id;
            let emp = new Employee(dep, per, e.position);
            emp.id = e.id;
            employees.push(emp);
        });
        return employees;
    }

    /**
     *
     * @param {any} entity
     * @returns {Promise<any>}
     */
    #remove(entity, schemaName) {
        let rep = this.connection.getRepository(schemaName);
        return rep.remove(entity);
    }

    /**
     *
     * @param {Department | Department[]} departments
     * @returns {Promise<Department | Departments[]>}
     */
    removeDepartments(departments) {
        return this.#remove(departments, Department.schemaName);
    }

    /**
     *
     * @param {Person | Person[]} persons
     * @returns {Promise<Person | Person[]>}
     */
    removePersons(persons) {
        return this.#remove(persons, Person.schemaName);
    }

    /**
     *
     * @param {Employee | Employee[]} employees
     * @returns {Promise<Employee | Employee[]>}
     */
    removeEmployees(employees) {
        return this.#remove(employees, Employee.schemaName);
    }
}

File: Lua/4DaysORM/src/Services/EntityService.lua
local Department = require "Models.Department"
local Person = require "Models.Person"
local Employee = require "Models.Employee"

local function get_by_id (entity_type)
    return function (id) return entity_type.get:where { id = id }:first() end
end

local function save (entity)
    entity:save ()
end

local function new (entity_type, fields)
    if type(fields) ~= "table" then error ("fields should be table, not " .. type (fields)) end
    local entity = entity_type(fields)
    save (entity)
end

local function all (entity_type)
    return function () return entity_type.get:all() end
end

local EntityService = {
    get_department = get_by_id (Department),
    get_person = get_by_id (Person),
    get_employee = get_by_id (Employee),

    add_new_department = function (name)
        new (Department, { name = name })
    end,

    add_new_person = function (name, address, birth_date)
        new (
            Person,
            {
                name = name,
                address = address,
                birthDate = birth_date
            }
        )
    end,

    add_new_employee = function (person, department, position)
        new (
            Employee,
            {
                personId = person.id,
                departmentId = department.id,
                position = position
            }
        )
    end,

    save = save,
    get_all_departments = all (Department),
    get_all_persons = all (Person),
    get_all_employees = all (Employee),

    remove_department = function (department)
        Employee.get:where { departmentId = department.id }:delete()
        department:delete()
    end,

    remove_person = function (person)
        Employee.get:where {personId = person.id }:delete()
        person:delete()
    end,

    remove_employee = function (employee)
        employee:delete()
    end,
}

function EntityService:add_new_employee_and_person(department, name, address, birth_date, position)
    if not self.add_new_person then error "use EntityService:add_new_employee_and_person(department, name, address, birth_date, position)" end
    self.add_new_person (name, address, birth_date)
    local per = Person.get:where{ name=name, address=address, birthDate=birth_date }:first()
    self.add_new_employee (per, department, position)
end

function EntityService:remove_departments (departments)
    for _, v in pairs (departments) do
        self.remove_department (v)
    end
end

function EntityService:remove_persons (persons)
    for _, v in pairs (persons) do
        self.remove_person (v)
    end
end

function EntityService:remove_employees (employees)
    for _, v in pairs (employees) do
        self.remove_employee (v)
    end
end

return EntityService
